query-planner-js/src/buildPlan.ts

```diff
@@ -1637,7 +1642,9 @@ function inputsForRequire(graph: QueryGraph, entityType: ObjectType, edge: Edge,
   fullSelectionSet.add(new FieldSelection(new Field(entityType.typenameField()!)));
   fullSelectionSet.mergeIn(edge.conditions!);
   if (includeKeyInputs) {
-    fullSelectionSet.mergeIn(additionalKeyEdgeForRequireEdge(graph, edge).conditions!);
+    const keyCondition = getLocallySatisfiableKey(graph, edge.head);
+    assert(keyCondition, () => `No key found for require on ${edge}`);
```

Comment:
Maybe a slightly better error message? Suggestion: "Use of @require mandates a @key and no key was found on ${edge}"